### ISSUE TITLE
Iterating over `CSV.parse_line` causes an explosion of threads/fibers

### ISSUE THREAD

**Issue by robbavey** [jruby/jruby#7346]
**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
`jruby 9.3.6.0 (2.6.8) 2022-06-27 7a2cbcd376 OpenJDK 64-Bit Server VM 17+35 on 17+35 +jit [x86_64-darwin]`

* Operating system and platform (e.g. `uname -a`)
`Darwin`

**Actual Behavior**

```
jruby-9.3.6.0 :001 > require 'csv'
 => true 
jruby-9.3.6.0 :002 > Thread.list
 => [#<Thread:0x1f1574c@Finalizer run>, #<Thread:0x16c9e378@main run>] 
jruby-9.3.6.0 :003 > CSV.parse_line('x')
 => ["x"] 
jruby-9.3.6.0 :004 > Thread.list
 => [#<Thread:0x66fff42f@Fiber thread for block at: uri:classloader:/jruby/kernel/enumerator.rb:62 sleep>, #<Thread:0x308be1a3@NonBlockingInputStreamThread run>, #<Thread:0x1f1574c@Finalizer run>, #<Thread:0x16c9e378@main run>] 
```

The `fiber thread for block at` thread has the following stacktrace:

```
"Fiber thread for block at: uri:classloader:/jruby/kernel/enumerator.rb:62" #28 daemon prio=5 os_prio=31 cpu=1.36ms elapsed=82.58s tid=0x00007fa1a5816a00 nid=0x9d0b waiting on condition  [0x0000700003750000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17/Native Method)
	- parking to wait for  <0x0000000404fff8d0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@17/LockSupport.java:341)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(java.base@17/AbstractQueuedSynchronizer.java:506)
	at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@17/ForkJoinPool.java:3463)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@17/ForkJoinPool.java:3434)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@17/AbstractQueuedSynchronizer.java:1623)
	at java.util.concurrent.ArrayBlockingQueue.take(java.base@17/ArrayBlockingQueue.java:420)
	at org.jruby.ext.fiber.FiberQueue$1.run(org.jruby.dist/FiberQueue.java:55)
	at org.jruby.ext.fiber.FiberQueue$1.run(org.jruby.dist/FiberQueue.java:52)
	at org.jruby.RubyThread.executeTask(org.jruby.dist/RubyThread.java:1712)
	at org.jruby.RubyThread.executeTaskBlocking(org.jruby.dist/RubyThread.java:1686)
	at org.jruby.ext.fiber.FiberQueue.pop(org.jruby.dist/FiberQueue.java:91)
	at org.jruby.ext.fiber.ThreadFiber.exchangeWithFiber(org.jruby.dist/ThreadFiber.java:109)
	at org.jruby.ext.fiber.ThreadFiber.yield(org.jruby.dist/ThreadFiber.java:222)
	at org.jruby.ext.fiber.ThreadFiber$INVOKER$s$yield.call(org.jruby.dist/ThreadFiber$INVOKER$s$yield.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(org.jruby.dist/JavaMethod.java:842)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(org.jruby.dist/CachingCallSite.java:329)
	at org.jruby.runtime.callsite.CachingCallSite.call(org.jruby.dist/CachingCallSite.java:87)
	at org.jruby.ir.instructions.CallBase.interpret(org.jruby.dist/CallBase.java:549)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(org.jruby.dist/InterpreterEngine.java:361)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:72)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(org.jruby.dist/Interpreter.java:116)
	at org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(org.jruby.dist/MixedModeIRBlockBody.java:136)
	at org.jruby.runtime.IRBlockBody.doYield(org.jruby.dist/IRBlockBody.java:170)
	at org.jruby.runtime.BlockBody.yield(org.jruby.dist/BlockBody.java:108)
	at org.jruby.runtime.Block.yield(org.jruby.dist/Block.java:188)
	at org.jruby.ir.runtime.IRRuntimeHelpers.yield(org.jruby.dist/IRRuntimeHelpers.java:493)
	at org.jruby.ir.instructions.YieldInstr.interpret(org.jruby.dist/YieldInstr.java:86)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.processOtherOp(org.jruby.dist/StartupInterpreterEngine.java:176)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:104)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(org.jruby.dist/InterpreterEngine.java:86)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(org.jruby.dist/MixedModeIRMethod.java:201)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(org.jruby.dist/MixedModeIRMethod.java:188)
	at org.jruby.runtime.callsite.RefinedCachingCallSite.call(org.jruby.dist/RefinedCachingCallSite.java:95)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(org.jruby.dist/InterpreterEngine.java:338)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:72)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(org.jruby.dist/Interpreter.java:116)
	at org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(org.jruby.dist/MixedModeIRBlockBody.java:136)
	at org.jruby.runtime.IRBlockBody.doYield(org.jruby.dist/IRBlockBody.java:170)
	at org.jruby.runtime.BlockBody.yield(org.jruby.dist/BlockBody.java:108)
	at org.jruby.runtime.Block.yield(org.jruby.dist/Block.java:188)
	at org.jruby.ir.runtime.IRRuntimeHelpers.yield(org.jruby.dist/IRRuntimeHelpers.java:493)
	at org.jruby.ir.instructions.YieldInstr.interpret(org.jruby.dist/YieldInstr.java:86)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.processOtherOp(org.jruby.dist/StartupInterpreterEngine.java:176)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:104)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(org.jruby.dist/Interpreter.java:116)
	at org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(org.jruby.dist/MixedModeIRBlockBody.java:136)
	at org.jruby.runtime.IRBlockBody.doYield(org.jruby.dist/IRBlockBody.java:170)
	at org.jruby.runtime.IRBlockBody.yieldSpecific(org.jruby.dist/IRBlockBody.java:99)
	at org.jruby.runtime.Block.yieldSpecific(org.jruby.dist/Block.java:166)
	at org.jruby.util.StringSupport.rbStrEnumerateLines(org.jruby.dist/StringSupport.java:2074)
	at org.jruby.util.StringSupport.rbStrEnumerateLines(org.jruby.dist/StringSupport.java:1958)
	at org.jruby.RubyString.each_line(org.jruby.dist/RubyString.java:5944)
	at org.jruby.RubyString$INVOKER$i$each_line.call(org.jruby.dist/RubyString$INVOKER$i$each_line.gen)
	at org.jruby.runtime.callsite.RefinedCachingCallSite.call(org.jruby.dist/RefinedCachingCallSite.java:95)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(org.jruby.dist/CachingCallSite.java:192)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(org.jruby.dist/InterpreterEngine.java:337)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:72)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(org.jruby.dist/InterpreterEngine.java:86)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(org.jruby.dist/MixedModeIRMethod.java:201)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(org.jruby.dist/MixedModeIRMethod.java:188)
	at org.jruby.runtime.callsite.RefinedCachingCallSite.call(org.jruby.dist/RefinedCachingCallSite.java:95)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(org.jruby.dist/CachingCallSite.java:192)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(org.jruby.dist/InterpreterEngine.java:337)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:72)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(org.jruby.dist/MixedModeIRMethod.java:128)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(org.jruby.dist/MixedModeIRMethod.java:115)
	at org.jruby.runtime.callsite.RefinedCachingCallSite.call(org.jruby.dist/RefinedCachingCallSite.java:47)
	at org.jruby.ir.instructions.CallBase.interpret(org.jruby.dist/CallBase.java:549)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(org.jruby.dist/InterpreterEngine.java:361)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:72)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(org.jruby.dist/MixedModeIRMethod.java:128)
	at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(org.jruby.dist/MixedModeIRMethod.java:115)
	at org.jruby.RubyClass.finvokeWithRefinements(org.jruby.dist/RubyClass.java:512)
	at org.jruby.RubyBasicObject.send(org.jruby.dist/RubyBasicObject.java:1733)
	at org.jruby.RubyBasicObject$INVOKER$i$send.call(org.jruby.dist/RubyBasicObject$INVOKER$i$send.gen)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(org.jruby.dist/CachingCallSite.java:329)
	at org.jruby.runtime.callsite.CachingCallSite.call(org.jruby.dist/CachingCallSite.java:87)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(org.jruby.dist/CachingCallSite.java:94)
	at org.jruby.ir.instructions.CallBase.interpret(org.jruby.dist/CallBase.java:546)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(org.jruby.dist/InterpreterEngine.java:361)
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(org.jruby.dist/StartupInterpreterEngine.java:72)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(org.jruby.dist/Interpreter.java:116)
	at org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(org.jruby.dist/MixedModeIRBlockBody.java:136)
	at org.jruby.runtime.IRBlockBody.yieldSpecific(org.jruby.dist/IRBlockBody.java:76)
	at org.jruby.runtime.Block.yieldSpecific(org.jruby.dist/Block.java:157)
	at org.jruby.ext.fiber.ThreadFiber.lambda$createThread$0(org.jruby.dist/ThreadFiber.java:302)
	at org.jruby.ext.fiber.ThreadFiber$$Lambda$971/0x00000008012632c8.run(org.jruby.dist/Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17/ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17/Thread.java:833)
```

Running rc removes the `fiber thread for block` thread:

```
java.lang.System.gc
 => nil 
jruby-9.3.6.0 :006 > Thread.list
 => [#<Thread:0x308be1a3@NonBlockingInputStreamThread run>, #<Thread:0x1f1574c@Finalizer run>, #<Thread:0x16c9e378@main run>] 
```

But now a java thread dump has the following:

```

"Ruby-0-Fiber-1" #28 daemon prio=5 os_prio=31 cpu=2.25ms elapsed=166.87s tid=0x00007fa1a5816a00 nid=0x9d0b waiting on condition  [0x0000700003753000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17/Native Method)
	- parking to wait for  <0x0000000400fa4008> (a java.util.concurrent.SynchronousQueue$TransferStack)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@17/LockSupport.java:252)
	at java.util.concurrent.SynchronousQueue$TransferStack.transfer(java.base@17/SynchronousQueue.java:401)
	at java.util.concurrent.SynchronousQueue.poll(java.base@17/SynchronousQueue.java:903)
	at java.util.concurrent.ThreadPoolExecutor.getTask(java.base@17/ThreadPoolExecutor.java:1061)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17/ThreadPoolExecutor.java:1122)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17/Thread.java:833)

```

Looping through `parse_line` causes a large number of these threads/fibers to be created:

```
jruby-9.3.6.0 :008 > 1000.times do CSV.parse_line('x') end
 => 1000 
jruby-9.3.6.0 :009 > Thread.list.count
 => 542 
```

Running `java.lang.System.gc` causes the same behaviour as shown above

This issue appears to be linked to https://github.com/jruby/jruby/issues/6004, perhaps being exposed by updated CSV parsing in `ruby-2.6`/`jruby-9.3`

Note that this does not happen on `jruby-9.2.20.1`, but does on `jruby-9.3.0.0`

Relates: https://github.com/elastic/logstash/issues/14498

**Comment by headius** [comment]
This is an unfortunate side effect of fibers in JRuby, due to the fact that we must back every fiber with a native JVM thread. I suspect that these fibers are coming from an Enumerator being used to `next` through lines.

@robbavey Perhaps you can confirm what code is causing the fibers to be created, presumably in CSV somewhere?

> Note that this does not happen on jruby-9.2.20.1, but does on jruby-9.3.0.0

I would assume this is because 9.2.20.1 is on an older version of CSV that does not use `Enumerator#next` or fibers to do this iteration. If so, we should be able to trace this to a specific change in CSV.

FWIW these threads do eventually collect, but if they are left dangling (e.g. due to an incomplete iteration) that collection must be triggered by an object finalizer. The finalizer is triggered by GC, and tries to forcibly kill the fiber thread so that it can terminate and get collected. This process can take a few GC cycles, and if there's not much "pressure" on the heap to collect the fibers can accumulate. Too many, and we will be unable to create new fiber threads, leading to the error in https://github.com/elastic/logstash/issues/14498.

Code can be modified to run enumerations to completion, so that the fiber thread completes its run, but for a filtering function that may be inefficient (presumably you don't necessarily want to iterate everything once you've found what you're looking for). 

I will try to determine where these fibers are being created and whether we can do anything to make them more easily collectable.

For the future, we may have a solution with JDK 19 and Project Loom's "virtual threads", for which we will ship support in 9.3.8.0, but that is seriously bleeding edge technology and only a "preview" in the JDK 19 release.

**Comment by headius** [comment]
This appears to be a known and fixed issue in CSV: https://github.com/ruby/csv/issues/120

The CSV issue here caused the stack trace to be broken, due to the fact that they used a single `Enumerator#next` call to grab exactly one line (the first and only line passed to `parse_line`) which triggered a Fiber to be created. Because this Enumerator never completed, its associated Fiber lingers.

The fixed code now just uses `first` to get the line and does not trigger a Fiber nor a thread to be created.

There may be other places in CSV that use fibers in a similar way, but I believe this issue can be fixed by upgrading CSV.

**Comment by headius** [comment]
Latest CSV appears to still be compatible with Ruby 2.5 and higher, so I see no reason we couldn't upgrade. CRuby 2.6 never moved past 3.0.9, which is before the fix above, so we will have to diverge from strict version matching here.

I will push a PR to update JRuby 9.3 to latest CSV.

**Comment by headius** [comment]
Pushed #7348 which will fix this for JRuby 9.3.8.0. JRuby master (9.4) is already on a newer CSV.

**Comment by robbavey** [comment]
Thank you for the quick turnaround - I've verified locally that upgrading the gem to `3.2.5` does indeed fix the issue.
